lib/trace/context-manager.js

```diff
@@ -0,0 +1,136 @@
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict'
+
+module.exports = ContextManager;
+var TraceContext = require('./trace-context');
+var DictionaryManager = require('../dictionary/dictionary-manager');
+var AgentConfig = require('../config');
+var NoopSpan = require('./noop-span');
+var NoopTraceContext = require('./noop-trace-context');
+var debug = require('debug')('skywalking-context-manager')
+
+const NOOP_TRACE_CONTEXT = new NoopTraceContext();
+const NOOP_SPAN = new NoopSpan(NOOP_TRACE_CONTEXT);
+
+/**
+ * @author zhang xin
+ */
+function ContextManager() {
+    this._activeTraceContext = undefined;
+    this._dictionaryManager = new DictionaryManager();
+}
+
+ContextManager.prototype.createSpan = function (spanOptions) {
+    if (!AgentConfig.getApplicationId()) {
+        debug("use the noop-span before the application has been registered.");
+        return NOOP_SPAN;
+    }
+
+    var traceContext = new TraceContext(this._activeTraceContext, spanOptions);
+    this.active(traceContext);
+    return traceContext.span();
+};
+
+ContextManager.prototype.inject = function (contextCarrier) {
+    if (!AgentConfig.getApplicationId()) {
+        return;
+    }
+
+    this._activeTraceContext.inject.apply(this._activeTraceContext, arguments);
+};
+
+ContextManager.prototype.extract = function (contextCarrier) {
+    if (!AgentConfig.getApplicationId()) {
+        return;
+    }
+
+    this._activeTraceContext.extract.apply(this._activeTraceContext, arguments);
+};
+
+ContextManager.prototype.finishSpan = function (span) {
+    var finishTraceContext = span.traceContext();
+    finishTraceContext.finish.apply(finishTraceContext, arguments);
+    this.active(finishTraceContext.parentTraceContext.apply(finishTraceContext, []));
+};
+
+ContextManager.prototype.active = function (traceContext) {
+    this._activeTraceContext = traceContext;
+};
+
+ContextManager.prototype.activeTraceContext = function (traceContext) {
+    return this._activeTraceContext;
+};
+
+ContextManager.prototype.createEntrySpan = function (operationName, contextCarrier) {
+    var spanOptions = {
+        spanType: "ENTRY"
+    }
+
+    this._dictionaryManager.findOperationName(operationName, function (key, value) {
+        spanOptions[key] = value;
+    });
+
+    var span = this.createSpan(spanOptions);
+
+    if (contextCarrier) {
+        span.traceContext().extract.apply(span.traceContext(), [contextCarrier, this._agentConfig]);
+    }
+
+    return span;
+};
+
+ContextManager.prototype.createExitSpan = function (operationName, peerId, contextCarrier) {
+    if (!AgentConfig.getApplicationId()) {
```

Comment:
When  application instance id is NULL, span shouldn't be created.

Reply:
And `#createSpan` has already included this check, where do you prefer to do this?